File: 647.palindromic-substrings.py
#
# @lc app=leetcode id=647 lang=python3
#
# [647] Palindromic Substrings
#

# @lc code=start
class Solution:
    # for each char, try to expand on both sides to form palindromes
    # T: O(n ^ 2)
    # S: O(1)
    def countSubstrings(self, s: str) -> int:
        self.ans = 0
        for i in range(len(s)):
            self.expand_palindrome(s, i, i)  # odd length
            self.expand_palindrome(s, i, i + 1)  # even length
        return self.ans

    def expand_palindrome(self, s: str, l: int, r: int) -> int:
        while l >= 0 and r < len(s) and s[l] == s[r]:
            self.ans += 1
            l -= 1
            r += 1

# @lc code=end

File: 647.palindromic-substrings.cpp
/*
 * @lc app=leetcode.cn id=647 lang=cpp
 *
 * [647] 回文子串
 */
#include <string>

using namespace std;

// @lc code=start
class Solution {
public:
    // Similar to Q5, but instead of finding the longest, count all palindromes
    // Two pointers solution
    // T: O(n ^ 2)
    // S: O(1)
    int countSubstrings(string s) {
        int n = s.size(), ans = 0;
        for (int i = 0; i < (n << 1) - 1; ++i) {
            // for each char, try to expand on both sides to form palindromes
            int l = i >> 1; 
            int r = l + (i & 1);
            while (l >= 0 && r < n && s[l] == s[r]) {
                --l;
                ++r;
                ++ans;
            }
        }
        return ans;
    }
};
// @lc code=end

